api/base/serializers.py

```diff
@@ -135,6 +137,90 @@ def to_representation(self, value):
         return {'links': {self.link_type: {'href': url, 'meta': meta}}}
 
 
+class JSONAPIHyperlinkedRelatedField(ser.HyperlinkedRelatedField):
+    """
+    HyperlinkedRelated field that returns a nested dict with url,
+    optional meta information, and link_type.
+
+    Example:
+
+        branched_from = JSONAPIHyperlinkedRelatedField(view_name='nodes:node-detail', lookup_field='pk',
+                                                    lookup_url_kwarg='node_id', read_only=True, link_type='related')
+
+    """
+    def __init__(self, view_name=None, **kwargs):
+        self.meta = kwargs.pop('meta', None)
+        self.link_type = kwargs.pop('link_type', 'url')
+        super(JSONAPIHyperlinkedRelatedField, self).__init__(view_name, **kwargs)
+
+    def to_representation(self, value):
+        """
+        Returns nested dictionary in format {'links': {'self.link_type': ... }
+
+        If no meta information, self.link_type is equal to a string containing link's URL.  Otherwise,
+        the link is represented as a links object with 'href' and 'meta' members.
+        """
+        url = super(JSONAPIHyperlinkedRelatedField, self).to_representation(value)
+
+        meta = {}
+        for key in self.meta or {}:
+            meta[key] = _rapply(self.meta[key], _url_val, obj=value, serializer=self.parent)
+        self.meta = meta
+
+        return {'links': {self.link_type: {'href': url, 'meta': self.meta}}}
+
+
+class JSONAPIHyperlinkedGuidRelatedField(ser.Field):
+    def __init__(self, **kwargs):
+        self.meta = kwargs.pop('meta', {})
+        self.link_type = kwargs.pop('link_type', 'url')
+        super(JSONAPIHyperlinkedGuidRelatedField, self).__init__(read_only=True, **kwargs)
+
+    def get_view_name(self, guid):
+        GUID_VIEWS = {
+            'node': {
+                'view_name': 'nodes:node-detail',
+                'lookup_field': 'pk',
+                'lookup_url_kwarg': 'node_id'
+            },
+            'comment': {
+                'view_name': 'comments:comment-detail',
+                'lookup_field': 'pk',
+                'lookup_url_kwarg': 'comment_id'
+            },
+        }
+
+        # get target type from guid
+        guid_object = Guid.load(guid).referent
+
+        if isinstance(guid_object, Node):
+            self.link_type = 'related'
+            guid_views = GUID_VIEWS['node']
+
+        elif isinstance(guid_object, Comment):
+            self.link_type = 'self'
+            guid_views = GUID_VIEWS['comment']
+
```

Comment:
This function should deal with cases where Guid is None, or of a type other than Node or Comment.